tensorflow/compiler/tf2tensorrt/trt_convert.cc

```diff
@@ -0,0 +1,389 @@
+    return errors::Internal(
+        "Multiple engines found, but we can only serialize one");
+  }
+  const std::unique_ptr<EngineContext>& engine =
+      resource->cache_.begin()->second;
+  if (!engine || !engine->cuda_engine) {
+    // TODO(tfeher): Engine could have failed to build. Handle that correctly.
+    return errors::Internal("Engine not found for", node.name());
+  }
+
+  // Serialize the engine.
+  engine_data->reset(engine->cuda_engine->serialize());
+
+  return Status::OK();
+}
+
+// Saves the TRT engines a static attribute of the TRTEngineOp nodes.
+Status ConvertToStaticEngine(const GraphDef graph_def,
+                             GraphDef* static_graph_def, Session* session) {
+  static_graph_def->CopyFrom(graph_def);
+  VLOG(1) << "Saving TRT engines as static engine";
+  std::string op{"TRTEngineOp"};
+  for (auto& node : *(static_graph_def->mutable_node())) {
+    if (!op.compare(node.op())) {
+      VLOG(2) << "Saving TRT engine for " << node.name()
+              << ", device: " << node.device();
+      TrtUniquePtrType<nvinfer1::IHostMemory> engine_data;
+      TF_RETURN_IF_ERROR(ReadSerializedEngine(node, session, &engine_data));
+      auto* attr = node.mutable_attr();
+      AttrValue static_engine;
+      static_engine.set_b(true);
+      AttrValue engine_string;
+      engine_string.set_s(engine_data->data(), engine_data->size());
+      (*attr)["static_engine"] = static_engine;
+      (*attr)["serialized_segment"] = engine_string;
+    }
+  }
+  return Status::OK();
+}
+
+Status ValidateConversionParams(const TfTrtConversionParams& p, int n_inputs) {
+  if (p.precision_mode == TrtPrecisionMode::INT8 && p.use_calibration) {
+    return errors::InvalidArgument(
+        "Calibration not yet implemented through the C++ interface. Please use "
+        "our Python API for calibration.");
+  }
+  if (p.convert_to_static_engine && n_inputs == 0) {
+    return errors::InvalidArgument(
+        "TRT Engine needs to be built before we can convert it to static "
+        "engine. Please provide input data to build the model.");
+  }
+  if (!p.convert_to_static_engine && n_inputs >= 0) {
+    // After the conversion, the session that was used to build the engines
+    // will be destroyed. If we do not convert the engine to static engine,
+    // then we loose the engines.
+    //
+    // TODO(tfeher): Provide a way to save dynamic engines and remove this
+    // warning.
+    LOG(WARNING)
+        << "Skipping build mode because we cannot save the "
+           "engines. Use convert_to_static_engines=true conversion "
+           "parameter to enable build mode and save the engines in the graph.";
+  }
+  if (!p.allow_build_at_runtime && n_inputs == 0) {
+    LOG(WARNING)
+        << "TRT will not be used since allow_build_at_runtime is disabled and "
+           "no inputs are provided to build during conversion.";
+  }
+  return Status::OK();
+}
+
+StatusOr<GraphDef> ConvertAndBuild(
```

Comment:
Only this routine needs to stay in this namespace, all the rest should be in "anonymous namespace", right?